src/MonC/Artifact.cs

```diff
@@ -0,0 +1,32 @@
+﻿using System;
+using System.IO;
+
+namespace MonC
+{
+    /// <summary>
+    /// Abstract compiler artifact which may potentially own unmanaged resources
+    /// </summary>
+    public abstract class Artifact : IDisposable
+    {
+        public virtual void Dispose()
+        {
+        }
+    }
+
+    /// <summary>
+    /// Abstract compiler artifact intended to be used as linker input
+    /// </summary>
+    public abstract class ModuleArtifact : Artifact
+    {
+        public virtual void WriteListing(TextWriter writer)
+        {
+        }
+    }
+
+    /// <summary>
+    /// Abstract compiler artifact intended to be used as VM input
+    /// </summary>
+    public abstract class ExecutableArtifact : Artifact
```

Comment:
I don't see this being used anywhere?

Reply:
This whole file is actually backported from the llvm branch where the new Driver is being implemented.

It's used here:
https://github.com/jackoalan/MonC/blob/llvm/src/Driver/JobAction.cs#L42